flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/AggregatedStatisticsSerializer.java

```diff
@@ -0,0 +1,175 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.sink.shuffle;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import org.apache.flink.api.common.typeutils.CompositeTypeSerializerSnapshot;
+import org.apache.flink.api.common.typeutils.TypeSerializer;
+import org.apache.flink.api.common.typeutils.TypeSerializerSnapshot;
+import org.apache.flink.api.common.typeutils.base.EnumSerializer;
+import org.apache.flink.api.common.typeutils.base.ListSerializer;
+import org.apache.flink.api.common.typeutils.base.LongSerializer;
+import org.apache.flink.api.common.typeutils.base.MapSerializer;
+import org.apache.flink.core.memory.DataInputView;
+import org.apache.flink.core.memory.DataOutputView;
+import org.apache.iceberg.SortKey;
+
+public class AggregatedStatisticsSerializer extends TypeSerializer<AggregatedStatistics> {
+  private final TypeSerializer<SortKey> sortKeySerializer;
+  private final EnumSerializer<StatisticsType> statisticsTypeSerializer;
+  private final MapSerializer<SortKey, Long> keyFrequencySerializer;
+  private final ListSerializer<SortKey> rangeBoundsSerializer;
+
+  AggregatedStatisticsSerializer(TypeSerializer<SortKey> sortKeySerializer) {
+    this.sortKeySerializer = sortKeySerializer;
+    this.statisticsTypeSerializer = new EnumSerializer<>(StatisticsType.class);
+    this.keyFrequencySerializer = new MapSerializer<>(sortKeySerializer, LongSerializer.INSTANCE);
+    this.rangeBoundsSerializer = new ListSerializer<>(sortKeySerializer);
+  }
+
+  @Override
+  public boolean isImmutableType() {
+    return false;
+  }
+
+  @Override
+  public TypeSerializer<AggregatedStatistics> duplicate() {
+    return new AggregatedStatisticsSerializer(sortKeySerializer);
+  }
+
+  @Override
+  public AggregatedStatistics createInstance() {
+    return new AggregatedStatistics(0, StatisticsType.Map, Collections.emptyMap(), null);
+  }
+
+  @Override
+  public AggregatedStatistics copy(AggregatedStatistics from) {
+    return new AggregatedStatistics(
+        from.checkpointId(), from.type(), from.keyFrequency(), from.rangeBounds());
+  }
+
+  @Override
+  public AggregatedStatistics copy(AggregatedStatistics from, AggregatedStatistics reuse) {
+    // no benefit of reuse
+    return copy(from);
+  }
+
+  @Override
+  public int getLength() {
+    return -1;
+  }
+
+  @Override
+  public void serialize(AggregatedStatistics record, DataOutputView target) throws IOException {
+    target.writeLong(record.checkpointId());
```

Comment:
Do we want to add a version?

Reply:
This implements the `TypeSerializer<T>` that is used by Flink runtime (like network and checkpoint state). All `TypeSerializer` implements `TypeSerializerSnapshot` to check compatibility and schema evolution story.

```
  public static class AggregatedStatisticsSerializerSnapshot
      extends CompositeTypeSerializerSnapshot<
          AggregatedStatistics, AggregatedStatisticsSerializer> {
```

this doesn't implement the `SimpleVersionedSerializer`, which is commonly used by the FLIP-27 source interface.

Reply:
Thanks for explaining!